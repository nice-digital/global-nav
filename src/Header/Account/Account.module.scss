@import '~@nice-digital/nds-core/scss/core';
@import '../header-settings';

.account {
  position: relative;
}

.button {
  background: #fff;
  border: 1px solid $nds-colour-text;
  color: $nds-colour-text;
  cursor: pointer;
  display: inline-block;
  font-family: inherit;
  font-size: inherit;
  font-weight: normal;
  height: rem($search-height);
  line-height: rem($search-height);
  padding: rem(0 $nds-spacing-small);
  text-decoration: none;
  vertical-align: middle;
  white-space: nowrap;

  &:focus {
    @include nds-default-focus-style;
  }

  &:visited {
    color: $nds-colour-text;
  }

  &:hover {
    border: 1px solid $nds-colour-text; // NICE.bootstrap removes border on hover by default
  }
}

.myAccount {
  &:after {
    border-left: rem($nds-spacing-x-small) solid transparent;
    border-right: rem($nds-spacing-x-small) solid transparent;
    border-top: rem($nds-spacing-small) solid $nds-colour-nice-dark-grey;
    content: '';
    display: inline-block;
    height: 0;
    margin-left: rem($nds-spacing-small);
    vertical-align: middle;
    width: 0;
  }

  &[aria-expanded='true']:after {
    border-bottom: rem($nds-spacing-small) solid $nds-colour-text;
    border-top: 0;
  }
}

.menu {
  background: $nds-colour-true-white;
  border: 1px solid $nds-colour-nice-dark-grey;
  list-style: none;
  margin: rem($nds-spacing-small + 2px 0 0 0);
  padding: 0;
  position: absolute;
  right: 0;
  top: 100%;
  z-index: $base-z-index;

  // We move focus when we expand via keyboard navigation
  &:focus {
    @include nds-default-focus-style;
  }

  &[aria-hidden='true'] {
    display: none;
  }

  &:before {
    border-bottom: rem($nds-spacing-small) solid $nds-colour-nice-dark-grey;
    border-left: rem($nds-spacing-x-small) solid transparent;
    border-right: rem($nds-spacing-x-small) solid transparent;
    bottom: 100%;
    content: '';
    display: block;
    height: 0;
    position: absolute;
    right: rem($nds-spacing-medium);
    width: 0;
  }

  &:after {
    border-bottom: rem($nds-spacing-small / 1.25) solid $nds-colour-true-white;
    border-left: rem($nds-spacing-x-small / 1.25) solid transparent;
    border-right: rem($nds-spacing-x-small / 1.25) solid transparent;
    bottom: 100%;
    content: '';
    display: block;
    height: 0;
    margin-right: 1px;
    position: absolute;
    right: rem($nds-spacing-medium);
    width: 0;
  }

  li {
    border-top: 1px solid $nds-colour-nice-grey-2;
    list-style: none;
    margin: 0;
  }

  a {
    color: $nds-colour-link;
    display: block;
    padding: rem($nds-spacing-small $nds-spacing-medium);
    text-decoration: none;
    white-space: nowrap;

    &:focus {
      @include nds-default-focus-style;
    }

    &:hover {
      background: $nds-colour-nice-grey-3;
      text-decoration: underline;
    }
  }

  span {
    padding: rem($nds-spacing-small $nds-spacing-medium);
}
