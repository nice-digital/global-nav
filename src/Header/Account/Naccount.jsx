import React, { useEffect, useState, useCallback } from "react";
import PropTypes from "prop-types";
import classnames from "classnames";

import {
	checkIsLoggedIn as niceAccountsLoggedIn,
	getDomainBaseUrl,
} from "./nice-accounts";
import styles from "./Account.module.scss";
import {
	trackEvent,
	defaultEventCategory,
	headerClickEventAction,
} from "../../tracker";

const escapeKeyCode = 27;

function Naccount(props) {
	// constructor(props) {
	// 	super(props);

	// state = {
	// 	isExpanded: false,
	// 	useIdAM: props.provider == Account.providers.idam,
	// };

	// 	this.handleMyAccountButtonClick =
	// 		this.handleMyAccountButtonClick;
	// 	this.handleKeyDown = this.handleKeyDown;
	// 	this.handleMenuItemClick = this.handleMenuItemClick;
	// 	this.handleMegaMenuClick = this.handleMegaMenuClick;
	// }

	// const [providers] = useState({
	// 	idam: "idam",
	// 	niceAccounts: "niceAccounts",
	// });

	const [state, setState] = useState({
		isExpanded: false,
		useIdAM: props.provider == Naccount.providers.idam,
	});

	const handleMyAccountButtonClick = useCallback((e) => {
		// debugger;
		setState(function (prevState) {
			return { isExpanded: !prevState.isExpanded };
		}, focusAccount(e, state.isExpanded));
		// if (!e.pageX && state.isExpanded) {
		// 	focusAccount();
		// }
	}, []);
	function focusAccount(e, expanded) {
		// debugger;
		console.log("focussing", e, " ", expanded);
		const accountMenu = document.getElementById("my-account");
		accountMenu.setAttribute("tabIndex", -1);
		accountMenu.focus();
	}

	useEffect(
		(a, b, c, d, e) => {
			console.log(a, b, c, d, e);
		},
		[state.isExpanded]
	);
	useEffect(() => {
		// console.log("use effect ran");
		// if (state.isExpanded) {
		// 	focusAccount();
		// }
	}, [state.isExpanded]);

	function handleKeyDown(event) {
		if (event.keyCode === escapeKeyCode) {
			event.preventDefault();
			setState({
				isExpanded: false,
			});
			document.getElementById("my-account-button").focus();
		}
	}

	// NOTE: We would benefit from managing the state higher up
	function handleMegaMenuClick(event) {
		let megaMenu = document.querySelector("#header-menu");

		if (megaMenu.contains(event.target)) {
			setState({
				isExpanded: false,
			});
			megaMenu.focus();
		}
	}

	function handleMenuItemClick(e) {
		const href = e.currentTarget.getAttribute("href");

		let eventLabel;

		if (href.indexOf("editprofile") > -1) {
			eventLabel = "Edit profile";
		} else if (href.indexOf("signout") > -1) {
			eventLabel = "Sign out";
		} else if (href.indexOf("signin") > -1) {
			eventLabel = "Sign in";
		}

		if (eventLabel) {
			e.preventDefault();

			trackEvent(
				defaultEventCategory,
				headerClickEventAction,
				eventLabel,
				null,
				href,
				function () {
					window.location.href = href;
				}
			);
		}
	}

	useEffect(() => {
		const consultationsResponsesLink = {
			"Consultation responses": "https://www.nice.org.uk/consultations/",
		};

		if (state.useIdAM) {
			//nice accounts supplies links like: {"John Holland":"https://accounts.nice.org.uk/users/143980/editprofile","Sign out":"https://accounts.nice.org.uk/signout"}
			//idam supplies links like:[{ key: "My profile", value: "/Account/todo" },{ key: "Sign out", value: "/Account/Logout" }]
			//the following just converts the idam format to the nice accounts format.

			const { displayName } = props;
			const isLoggedIn = !!displayName;

			let links = props.links.reduce(function (links, link) {
				links[link.text] = link.url;
				return links;
			}, {});

			if (isLoggedIn) {
				links = { ...consultationsResponsesLink, ...links };
			}

			const convertedData = {
				display_name: displayName,
				links: links,
			};

			if (props.onLoginStatusChecked) {
				props.onLoginStatusChecked(convertedData);
			}
		} else {
			//NICE accounts
			niceAccountsLoggedIn(props.environment)
				.then(function (data) {
					if (props.onLoginStatusChecked) {
						data.links = { ...consultationsResponsesLink, ...data.links };
						props.onLoginStatusChecked(data);
					}
				})
				.catch(function (e) {
					console.warn("Couldn't load account data from NICE accounts", e);
				});
		}

		document.addEventListener("click", handleMegaMenuClick);
	}, []);

	const { accountsData, environment } = props;

	let signInLink = {};
	if (state.useIdAM) {
		signInLink = props.links[0];
	} else {
		signInLink["text"] = "Sign in";
		signInLink["url"] = getDomainBaseUrl(environment) + "signin";
	}

	return props.isLoggedIn ? (
		<div className={styles.account}>
			<button
				className={classnames(styles.button, styles.myAccount)}
				id="my-account-button"
				aria-controls="my-account"
				aria-haspopup="menu"
				aria-expanded={state.isExpanded}
				onClick={handleMyAccountButtonClick}
				onKeyDown={handleKeyDown}
			>
				My account
			</button>
			<ul
				className={styles.menu}
				id="my-account"
				role="menu"
				aria-hidden={!state.isExpanded}
				aria-labelledby="my-account-button"
				onKeyDown={handleKeyDown}
			>
				{accountsData.links &&
					Object.keys(accountsData.links).map(function (text, i) {
						return (
							<li key={i} role="presentation">
								<a
									href={accountsData.links[text]}
									role="menuitem"
									onClick={handleMenuItemClick}
									onKeyDown={handleKeyDown}
									data-hj-suppress={
										accountsData.links[text].indexOf("profile") > -1 ? "" : null
									}
								>
									{text}
								</a>
							</li>
						);
					})}
			</ul>
		</div>
	) : (
		<a
			href={signInLink.url}
			className={styles.button}
			onClick={handleMenuItemClick}
		>
			{signInLink.text}
		</a>
	);
}

export default Naccount;

Naccount.providers = {
	idam: "idam",
	niceAccounts: "niceAccounts",
};

Naccount.propTypes = {
	isLoggedIn: PropTypes.bool.isRequired,
	onLoginStatusChecked: PropTypes.func,
	accountsData: PropTypes.shape({
		display_name: PropTypes.string,
		thumbnail: PropTypes.string,
		links: PropTypes.object,
	}),
	environment: PropTypes.oneOf(["live", "test", "beta", "local"]),
	provider: PropTypes.oneOf([
		Naccount.providers.niceAccounts,
		Naccount.providers.idam,
	]),
	links: PropTypes.arrayOf(
		PropTypes.shape({
			text: PropTypes.string.isRequired,
			url: PropTypes.string.isRequired,
		})
	),
	displayName: PropTypes.string,
};

Naccount.defaultProps = {
	environment: "live",
	provider: "niceAccounts",
};
