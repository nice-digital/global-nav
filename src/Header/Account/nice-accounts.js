// Checks if you are logged in via NICE Accounts
// Returns a promise that resolves with the data from NICE Accounts.
// Returns a promise that rejects if the data could not be loaded.

export const checkIsLoggedIn = function (environment) {
	return new Promise(function (resolve, reject) {
		const url = getDomainBaseUrl(environment) + "tophat";

		var body = document.body;
		var script = document.createElement("script");
		script.src =
			url +
			(~url.indexOf("?") ? "&" : "?") +
			Math.floor(Math.random() * 10000000000);

		let done = false;

		const timeout = setTimeout(() => {
			if (!done) {
				done = true;
				if (body && script.parentNode) {
					script.onload = script.onreadystatechange = null;
					body.removeChild(script);
				}
				reject(new Error("Timeout loading NICE Accounts tophat script"));
			}
		}, 5000); // 5 seconds

		script.onload = script.onreadystatechange = function () {
			if (
				!done &&
				(!script.readyState ||
					script.readyState === "loaded" ||
					script.readyState === "complete")
			) {
				done = true;
				clearTimeout(timeout);
				// Handle memory leak in IE
				script.onload = script.onreadystatechange = null;
				if (body && script.parentNode) {
					body.removeChild(script);
				}

				resolve(window._na);
			}
		};
		script.onerror = function () {
			if (!done) {
				done = true;
				clearTimeout(timeout);
				if (body && script.parentNode) {
					body.removeChild(script);
				}
				reject(new Error("Failed to load NICE Accounts tophat script"));
			}
		};

		body.insertBefore(script, body.firstChild);
	});
};

export const getDomainBaseUrl = function (environment) {
	return `https://${
		environment !== "live" ? environment + "-" : ""
	}accounts.nice.org.uk/`;
};
