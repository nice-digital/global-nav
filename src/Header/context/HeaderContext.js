import React, { useEffect, useState, createContext } from "react";
import PropTypes from "prop-types";

const defaultValues = { accountMenuIsExpanded: false, idOfOpenDropdown: null };

export const HeaderContext = createContext(defaultValues);

export const HeaderContextProvider = function ({ children }) {
	const [idOfOpenDropdown, setidOfOpenDropdown] = useState(null);
	const [accountMenuIsExpanded, setAccountMenuIsExpanded] = useState(false);

	const value = {
		idOfOpenDropdown,
		setidOfOpenDropdown,
		accountMenuIsExpanded,
		setAccountMenuIsExpanded,
	};

	const locationOrHashChangeHandler = () => {
		console.log("closing any open dropdown due to location or hash change");
		setidOfOpenDropdown(null);
	};

	useEffect(() => {
		window.addEventListener("hashchange", locationOrHashChangeHandler);
		return () => {
			window.removeEventListener("hashchange", locationOrHashChangeHandler);
		};
	}, []);

	useEffect(() => {
		locationOrHashChangeHandler();
	}, [location]);

	useEffect(() => {
		// we know both are now open
		if (accountMenuIsExpanded && idOfOpenDropdown) {
			setAccountMenuIsExpanded((prevState) => {
				// if prevstate is not null, you know it was already open, therefore close it
				if (prevState) return false;
			});
		}
	}, [accountMenuIsExpanded, idOfOpenDropdown]);

	return (
		<HeaderContext.Provider value={value}>{children}</HeaderContext.Provider>
	);
};

HeaderContextProvider.propTypes = {
	children: PropTypes.node,
};
