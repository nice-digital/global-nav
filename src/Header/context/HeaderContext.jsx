import {
	useEffect,
	useState,
	createContext,
	createRef,
	useCallback,
} from "react";

import PropTypes from "prop-types";

const defaultValues = { accountMenuIsExpanded: false, idOfOpenDropdown: null };

export const HeaderContext = createContext(defaultValues);

export const HeaderContextProvider = function ({ children }) {
	const [idOfOpenDropdown, setidOfOpenDropdown] = useState(null);
	const [accountMenuIsExpanded, setAccountMenuIsExpanded] = useState(false);
	const clickOutsideRef = createRef();

	const value = {
		idOfOpenDropdown,
		setidOfOpenDropdown,
		accountMenuIsExpanded,
		setAccountMenuIsExpanded,
		clickOutsideRef,
	};

	const handleClickOutsideHeaderMenu = useCallback(
		(e) => {
			if (
				clickOutsideRef.current &&
				clickOutsideRef.current.contains(e.target) == false
			) {
				setidOfOpenDropdown(null);
			}
		},
		[clickOutsideRef]
	);

	useEffect(() => {
		// listen for clicks outside of header menu to close dropdown
		document.body.addEventListener("click", handleClickOutsideHeaderMenu);
		return () => {
			document.body.removeEventListener("click", handleClickOutsideHeaderMenu);
		};
	}, [handleClickOutsideHeaderMenu]);

	const locationHref = typeof location == "undefined" ? null : location.href;

	useEffect(() => {
		setidOfOpenDropdown(null);
	}, [locationHref]);

	const popStateHandler = () => {
		setidOfOpenDropdown(null);
	};

	useEffect(() => {
		window && window.addEventListener("popstate", popStateHandler);
		return () => {
			window && window.removeEventListener("popstate", popStateHandler);
		};
	}, []);

	useEffect(() => {
		// we know both are now open
		if (accountMenuIsExpanded && idOfOpenDropdown) {
			setAccountMenuIsExpanded((prevState) => {
				// if prevstate is not null, you know it was already open, therefore close it
				if (prevState) return false;
			});
		}
	}, [accountMenuIsExpanded, idOfOpenDropdown]);

	return (
		<HeaderContext.Provider value={value}>{children}</HeaderContext.Provider>
	);
};

HeaderContextProvider.propTypes = {
	children: PropTypes.node,
};
