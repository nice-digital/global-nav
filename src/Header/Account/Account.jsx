import React, {
	useContext,
	useEffect,
	useState,
	useCallback,
	useRef,
} from "react";
import { HeaderContext } from "../context/HeaderContext";
import PropTypes from "prop-types";
import classnames from "classnames";

import {
	checkIsLoggedIn as niceAccountsLoggedIn,
	getDomainBaseUrl,
} from "./nice-accounts";
import styles from "./Account.module.scss";
import {
	trackEvent,
	defaultEventCategory,
	headerClickEventAction,
} from "../../tracker";

function Account(props) {
	// We've left this is state as per the pre-hook implementation
	const [doesUseIdAM] = useState(props.provider == Account.providers.idam);

	const { accountMenuIsExpanded, setAccountMenuIsExpanded } =
		useContext(HeaderContext);

	const keypress = useRef(),
		myAccountButton = useRef(),
		myAccountMenu = useRef();

	const handleMyAccountButtonClick = useCallback((e) => {
		keypress.current = e.pageX;
		setAccountMenuIsExpanded((prevState) => !prevState);
	}, []);

	function focusAccount() {
		myAccountMenu && myAccountMenu.current.setAttribute("tabIndex", -1);
		myAccountMenu && myAccountMenu.current.focus();
	}

	useEffect(() => {
		if (!keypress.current && accountMenuIsExpanded) {
			focusAccount();
		}
	}, [accountMenuIsExpanded]);

	function handleKeyDown(e) {
		if (e.key === "Escape") {
			e.preventDefault();
			setAccountMenuIsExpanded(false);
			myAccountButton && myAccountButton.current.focus();
		}
	}

	function handleMenuItemClick(e) {
		const href = e.currentTarget.getAttribute("href");

		let eventLabel;

		if (href.indexOf("editprofile") > -1) {
			eventLabel = "Edit profile";
		} else if (href.indexOf("signout") > -1) {
			eventLabel = "Sign out";
		} else if (href.indexOf("signin") > -1) {
			eventLabel = "Sign in";
		}

		if (eventLabel) {
			e.preventDefault();

			trackEvent(
				defaultEventCategory,
				headerClickEventAction,
				eventLabel,
				null,
				href,
				function () {
					window.location.href = href;
				}
			);
		}
	}

	useEffect(() => {
		const consultationsResponsesLink = {
			"Consultation responses": "https://www.nice.org.uk/consultations/",
		};

		if (doesUseIdAM) {
			//nice accounts supplies links like: {"John Holland":"https://accounts.nice.org.uk/users/143980/editprofile","Sign out":"https://accounts.nice.org.uk/signout"}
			//idam supplies links like:[{ key: "My profile", value: "/Account/todo" },{ key: "Sign out", value: "/Account/Logout" }]
			//the following just converts the idam format to the nice accounts format.

			const { displayName } = props;
			const isLoggedIn = !!displayName;

			let links = props.links.reduce(function (links, link) {
				links[link.text] = link.url;
				return links;
			}, {});

			if (isLoggedIn) {
				links = { ...consultationsResponsesLink, ...links };
			}

			const convertedData = {
				display_name: displayName,
				links: links,
			};

			if (props.onLoginStatusChecked) {
				props.onLoginStatusChecked(convertedData);
			}
		} else {
			//NICE accounts
			niceAccountsLoggedIn(props.environment)
				.then(function (data) {
					if (props.onLoginStatusChecked) {
						data.links = { ...consultationsResponsesLink, ...data.links };
						props.onLoginStatusChecked(data);
					}
				})
				.catch(function (e) {
					console.warn("Couldn't load account data from NICE accounts", e);
				});
		}
	}, []);

	const { accountsData, environment } = props;

	let signInLink = {};
	if (doesUseIdAM) {
		signInLink = props.links[0];
	} else {
		signInLink["text"] = "Sign in";
		signInLink["url"] = getDomainBaseUrl(environment) + "signin";
	}

	return props.isLoggedIn ? (
		<div className={styles.account} data-tracking="Account panel">
			<button
				className={classnames(styles.button, styles.myAccount)}
				id="my-account-button"
				aria-controls="my-account"
				aria-haspopup="menu"
				aria-expanded={accountMenuIsExpanded}
				onClick={handleMyAccountButtonClick}
				onKeyDown={handleKeyDown}
				ref={myAccountButton}
			>
				My account
			</button>
			<ul
				className={styles.menu}
				id="my-account"
				role="menu"
				aria-hidden={!accountMenuIsExpanded}
				aria-labelledby="my-account-button"
				onKeyDown={handleKeyDown}
				ref={myAccountMenu}
			>
				{accountsData.links &&
					Object.keys(accountsData.links).map(function (text, i) {
						return (
							<li key={i} role="presentation">
								<a
									href={accountsData.links[text]}
									role="menuitem"
									onClick={handleMenuItemClick}
									onKeyDown={handleKeyDown}
									data-hj-suppress={
										accountsData.links[text].indexOf("profile") > -1 ? "" : null
									}
								>
									{text}
								</a>
							</li>
						);
					})}
			</ul>
		</div>
	) : (
		<a
			data-tracking="Sign in"
			href={signInLink.url}
			className={styles.button}
			onClick={handleMenuItemClick}
		>
			{signInLink.text}
		</a>
	);
}

export default Account;

Account.providers = {
	idam: "idam",
	niceAccounts: "niceAccounts",
};

Account.propTypes = {
	isLoggedIn: PropTypes.bool.isRequired,
	onLoginStatusChecked: PropTypes.func,
	accountsData: PropTypes.shape({
		display_name: PropTypes.string,
		thumbnail: PropTypes.string,
		links: PropTypes.object,
	}),
	environment: PropTypes.oneOf(["live", "test", "beta", "local"]),
	provider: PropTypes.oneOf([
		Account.providers.niceAccounts,
		Account.providers.idam,
	]),
	links: PropTypes.arrayOf(
		PropTypes.shape({
			text: PropTypes.string.isRequired,
			url: PropTypes.string.isRequired,
		})
	),
	displayName: PropTypes.string,
};

Account.defaultProps = {
	environment: "live",
	provider: "niceAccounts",
};
