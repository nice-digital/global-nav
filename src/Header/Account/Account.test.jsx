import React from "react";
import Account from "./Account";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import { GlobalNavContextProvider } from "./../context/HeaderContext";

import {
	eventName,
	defaultEventCategory,
	headerClickEventAction,
	eventTimeout,
} from "../../tracker";

jest.mock("./nice-accounts", () => ({
	checkIsLoggedIn: jest.fn(() =>
		Promise.resolve({
			links: {
				test: true,
			},
			test: true,
		})
	),
	getDomainBaseUrl: jest.requireActual("./nice-accounts.js").getDomainBaseUrl,
}));

const idamProps = {
	provider: "idam",
	links: [{ text: "sign in", url: "/signin" }],
};

describe("Account", () => {
	const accountsData = {
		display_name: "Joe Bloggs",
		links: {
			"Joe Bloggs": "https://accounts.nice.org.uk/users/12345/editprofile",
			"Sign out": "https://accounts.nice.org.uk/signout",
		},
	};

	afterEach(() => {
		document.body.innerHTML = "";
	});

	it("Renders without crashing", () => {
		const wrapper = shallow(<Account isLoggedIn={false} />);
		expect(wrapper).toHaveLength(1);
	});

	it("Matches snapshot when logged out", () => {
		const wrapper = shallow(<Account isLoggedIn={false} />);
		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it("Matches snapshot when logged in", () => {
		const wrapper = shallow(
			<Account isLoggedIn={true} accountsData={accountsData} />
		);

		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it("Matches snapshot when logged out using IDAM", () => {
		const wrapper = shallow(<Account isLoggedIn={false} {...idamProps} />);
		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it("Matches snapshot when logged in using IDAM", () => {
		const wrapper = shallow(
			<Account isLoggedIn={true} accountsData={accountsData} {...idamProps} />
		);
		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it("Is the correct authentication environment when supplied", () => {
		const wrapper = shallow(<Account isLoggedIn={false} environment="beta" />);

		expect(wrapper.find(".button").props().href).toBe(
			"https://beta-accounts.nice.org.uk/signin"
		);
	});

	it("Calls onLoginStatusChecked callback prop when mounted", (done) => {
		const onLoginStatusChecked = jest.fn();

		mount(
			<GlobalNavContextProvider>
				<Account
					isLoggedIn={false}
					onLoginStatusChecked={onLoginStatusChecked}
				/>
			</GlobalNavContextProvider>
		);

		setImmediate(() => {
			expect(onLoginStatusChecked).toHaveBeenCalledWith({
				links: {
					"Consultation responses": "https://www.nice.org.uk/consultations/",
					test: true,
				},
				test: true,
			});
			done();
		});
	});

	it("Expands account menu when button is clicked", () => {
		var appContainer = document.createElement("div");
		document.body.appendChild(appContainer);

		const wrapper = mount(
			<GlobalNavContextProvider>
				<Account isLoggedIn={true} accountsData={accountsData} />
			</GlobalNavContextProvider>,
			{ attachTo: appContainer }
		);

		wrapper.find("#my-account-button").simulate("click", { pageX: 99 });
		// wrapper.instance().forceUpdate();
		wrapper.update();

		expect(wrapper.find("#my-account-button").props()["aria-expanded"]).toBe(
			true
		);
		var menu = wrapper.find("#my-account");
		expect(menu.props()["aria-hidden"]).toBe(false);
		expect(menu.props()["tabIndex"]).toBe(undefined);
	});

	it("Moved focus to menu when enter key is pressed on collapsed button", () => {
		var appContainer = document.createElement("div");
		document.body.appendChild(appContainer);

		const wrapper = mount(
			<GlobalNavContextProvider>
				<Account isLoggedIn={true} accountsData={accountsData} />
			</GlobalNavContextProvider>,
			{ attachTo: appContainer }
		);

		wrapper.find("#my-account-button").simulate("click");
		wrapper.update();

		var menuElement = wrapper.find("#my-account").getDOMNode();
		expect(menuElement.getAttribute("tabIndex")).toEqual("-1");
		expect(menuElement).toBe(document.activeElement);
	});

	it("Collapses menu when escape key is pressed on button", () => {
		var appContainer = document.createElement("div");
		document.body.appendChild(appContainer);

		const wrapper = mount(
			<GlobalNavContextProvider>
				<Account isLoggedIn={true} accountsData={accountsData} />
			</GlobalNavContextProvider>,
			{ attachTo: appContainer }
		);

		wrapper.find("#my-account-button").simulate("click", {});
		wrapper.find("#my-account-button").simulate("keydown", { key: "Escape" });
		wrapper.update();
		expect(wrapper.find("#my-account-button").props()["aria-expanded"]).toBe(
			false
		);
		expect(wrapper.find("#my-account").props()["aria-hidden"]).toBe(true);
	});

	it("Collapses menu when escape key is pressed on link", () => {
		var appContainer = document.createElement("div");
		document.body.appendChild(appContainer);

		const wrapper = mount(
			<GlobalNavContextProvider>
				<Account isLoggedIn={true} accountsData={accountsData} />
			</GlobalNavContextProvider>,
			{ attachTo: appContainer }
		);

		wrapper.find("#my-account-button").simulate("click", {});

		wrapper
			.find("a[role='menuitem']")
			.first()
			.simulate("keydown", { key: "Escape" });

		expect(
			wrapper.find("[aria-controls='my-account']").props()["aria-expanded"]
		).toBe(false);
		expect(wrapper.find("#my-account").props()["aria-hidden"]).toBe(true);
	});

	describe("tracking", () => {
		beforeEach(() => {
			window.dataLayer = [];
		});

		afterAll(() => {
			// Cleanup
			delete window.dataLayer;
		});

		const trackingTest = (accountsData, linkText, eventLabel, href) => {
			const wrapper = shallow(
				<Account
					isLoggedIn={accountsData != null}
					accountsData={accountsData}
				/>
			);

			const preventDefault = jest.fn();

			const button = wrapper.find(`a[children="${linkText}"]`);

			button.props().onClick({
				preventDefault: preventDefault,
				currentTarget: {
					getAttribute: () => href,
				},
			});

			expect(window.dataLayer).toEqual([
				{
					event: eventName,
					eventCategory: defaultEventCategory,
					eventAction: headerClickEventAction,
					eventLabel: eventLabel,
					destinationUrl: href,
					eventCallback: expect.any(Function),
					eventTimeout: eventTimeout,
				},
			]);

			expect(preventDefault).toHaveBeenCalled();

			window.dataLayer[0].eventCallback();
			expect(window.location.href).toEqual(href);
		};

		it("should not send dataLayer event or prevent default for admin link click", () => {
			const wrapper = mount(
				<GlobalNavContextProvider>
					<Account
						isLoggedIn={true}
						accountsData={{
							display_name: "Joe Bloggs",
							links: {
								Admin: "https://accounts.nice.org.uk/admin",
							},
						}}
					/>
				</GlobalNavContextProvider>
			);

			const preventDefault = jest.fn();

			wrapper.find(`a[children="Admin"]`).simulate("click", {
				preventDefault: preventDefault,
				currentTarget: {
					getAttribute: () => "https://accounts.nice.org.uk/admin",
				},
			});

			expect(preventDefault).not.toHaveBeenCalled();

			expect(window.dataLayer).toHaveLength(0);
		});

		it("should send event to the dataLayer with navigating event callback for sign in button click", () => {
			trackingTest(
				null,
				"Sign in",
				"Sign in",
				"https://beta-accounts.nice.org.uk/signin"
			);
		});

		it("should send event to the dataLayer with navigating event callback for edit profile link click", () => {
			trackingTest(
				accountsData,
				"Joe Bloggs",
				"Edit profile",
				"https://accounts.nice.org.uk/users/12345/editprofile"
			);
		});

		it("should send event to the dataLayer with navigating event callback for sign out link click", () => {
			trackingTest(
				accountsData,
				"Sign out",
				"Sign out",
				"https://beta-accounts.nice.org.uk/signout"
			);
		});
	});
});
