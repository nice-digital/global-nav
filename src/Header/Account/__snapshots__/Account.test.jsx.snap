// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account Matches snapshot when logged in  using IDAM 1`] = `
<div
  className="account"
>
  <button
    aria-controls="my-account"
    aria-expanded={false}
    aria-haspopup="menu"
    className="button myAccount"
    id="my-account-button"
    onClick={[Function]}
    onKeyDown={[Function]}
  >
    My account
  </button>
  <ul
    aria-hidden={true}
    aria-labelledby="my-account-button"
    className="menu"
    id="my-account"
    onKeyDown={[Function]}
    role="menu"
  >
    <li
      key="0"
      role="presentation"
    >
      <a
        data-hj-suppress=""
        href="https://accounts.nice.org.uk/users/12345/editprofile"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
      >
        Joe Bloggs
      </a>
    </li>
    <li
      key="1"
      role="presentation"
    >
      <a
        data-hj-suppress={null}
        href="https://accounts.nice.org.uk/signout"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
      >
        Sign out
      </a>
    </li>
  </ul>
</div>
`;

exports[`Account Matches snapshot when logged in 1`] = `
<div
  className="account"
>
  <button
    aria-controls="my-account"
    aria-expanded={false}
    aria-haspopup="menu"
    className="button myAccount"
    id="my-account-button"
    onClick={[Function]}
    onKeyDown={[Function]}
  >
    My account
  </button>
  <ul
    aria-hidden={true}
    aria-labelledby="my-account-button"
    className="menu"
    id="my-account"
    onKeyDown={[Function]}
    role="menu"
  >
    <li
      key="0"
      role="presentation"
    >
      <a
        data-hj-suppress=""
        href="https://accounts.nice.org.uk/users/12345/editprofile"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
      >
        Joe Bloggs
      </a>
    </li>
    <li
      key="1"
      role="presentation"
    >
      <a
        data-hj-suppress={null}
        href="https://accounts.nice.org.uk/signout"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="menuitem"
      >
        Sign out
      </a>
    </li>
  </ul>
</div>
`;

exports[`Account Matches snapshot when logged out 1`] = `
<Account
  environment="live"
  isLoggedIn={false}
  provider="niceAccounts"
>
  <a
    className="button"
    href="https://accounts.nice.org.uk/signin"
    onClick={[Function]}
  >
    Sign in
  </a>
</Account>
`;

exports[`Account Matches snapshot when logged out using IDAM 1`] = `
<a
  className="button"
  href="/signin"
  onClick={[Function]}
>
  sign in
</a>
`;
